pkg/dataplane/util/image_registry.go

```diff
@@ -0,0 +1,127 @@
+package util
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"strings"
+
+	mc "github.com/openshift/api/machineconfiguration/v1"
+	ocpimage "github.com/openshift/api/operator/v1alpha1"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// machineConfigIgnition - holds the relevant parts of the ignition file that we need to create
+// the registries.conf on the dataplane nodes.
+type machineConfigIgnition struct {
+	Ignition struct {
+		Version string `json:"version"`
+	} `json:"ignition"`
+	Storage struct {
+		Files []struct {
+			Contents struct {
+				Compression string `json:"compression,omitempty"`
+				Source      string `json:"source"`
+			} `json:"contents"`
+			Mode      int    `json:"mode,omitempty"`
+			Overwrite bool   `json:"overwrite,omitempty"`
+			Path      string `json:"path"`
+		} `json:"files"`
+	} `json:"storage"`
+}
+
+// IsDisconnectedOCP - Will retrieve a ImageContentSourcePolicyList. If the list is not
+// empty, we can infer that the OCP cluster is a disconnected deployment.
+func IsDisconnectedOCP(ctx context.Context, helper *helper.Helper) (bool, error) {
+	icspList := ocpimage.ImageContentSourcePolicyList{}
+
+	listOpts := []client.ListOption{}
+	err := helper.GetClient().List(ctx, &icspList, listOpts...)
+	if err != nil {
+		return false, err
+	}
+
+	if len(icspList.Items) != 0 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// GetMCRegistryConf - will unmarshal the MachineConfig ignition file the machineConfigIgnition object.
+// This is then parsed and the base64 decoded string is returned.
+func GetMCRegistryConf(ctx context.Context, helper *helper.Helper) (string, error) {
+	var registriesConf string
+
+	masterMachineConfig, err := getMachineConfig(ctx, helper)
+	if err != nil {
+		return registriesConf, err
+	}
+
+	config := machineConfigIgnition{}
+	registriesConf, err = config.formatRegistriesConfString(&masterMachineConfig)
+	if err != nil {
+		return registriesConf, err
+	}
+
+	return registriesConf, nil
+}
+
+func (mci *machineConfigIgnition) removePrefixFromB64String() string {
+	const b64Prefix string = "data:text/plain;charset=utf-8;base64,"
+	if strings.HasPrefix(mci.Storage.Files[0].Contents.Source, b64Prefix) {
+		return mci.Storage.Files[0].Contents.Source[len(b64Prefix):]
+	}
+
+	return ""
```

Comment:
hm, this feels wrong, if it does not have a prefix then it is either an error or we can use the unprefixed value but returning empty here feels wrong. Or do we rely on the caller to error out if it gets an empty string?

Reply:
Yeah, I don't really like this in general tbh. I thought there would be a better way to parse that text based on the prefix. I might need to dig deeper into how MCO uses this value and see if there's some ideas for a cleaner implementation here. I can't really see it being defined as anything else, unless the Spec for this base64 string designation changes, but I guess we should return an error here. Or at least some feedback to the user that we couldn't retrieve the OCP `registries.conf` and we will fall back to the original implementation in edpm-ansible in those situations.

Reply:
Changed this to return an error if it can't find the prefix string in the output.